tests/query_test.py

```diff
@@ -66,19 +68,72 @@ def test_quote(self):
     quoted = query.let({"x": 1}, query.var("x"))
     assert self._q(query.quote(quoted)) == quoted
 
+  def test_lambda_query(self):
+    assert query.lambda_query(lambda a: query.add(a, a)) == {
+      "lambda": "auto0", "expr": {"add": ({"var": "auto0"}, {"var": "auto0"})}
+    }
+
+    expected = query.lambda_query(
+      lambda a: query.lambda_query(
+        lambda b: query.lambda_query(
+          lambda c: [a, b, c])))
+    assert expected == {
+      "lambda": "auto0",
+      "expr": {
+        "lambda": "auto1",
+        "expr": {
+          "lambda": "auto2",
+          "expr": [{"var": "auto0"}, {"var": "auto1"}, {"var": "auto2"}]
+        }
+      }
+    }
+
+    # Error in lambda should not affect future queries.
+    with self.assertRaises(Exception):
+      def fail():
+        raise Exception("foo")
+      query.lambda_query(lambda a: fail())
+    assert query.lambda_query(lambda a: a) == {"lambda": "auto0", "expr": {"var": "auto0"}}
+
+  def test_lambda_query_multithreaded(self):
+    """Test that lambda_query works in simultaneous threads."""
+    events = []
+
+    def do_a():
+      def do_lambda(a):
+        events.append(0)
+        sleep(1)
```

Comment:
Not a fan of this, but the ordering would be wrong if it ran out of order anyway. I think this is good enough for what's being tested here.